hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/ConnectionManager.java

```diff
@@ -203,9 +210,16 @@ public ConnectionContext getConnection(UserGroupInformation ugi,
       try {
         pool = this.pools.get(connectionId);
         if (pool == null) {
+          if(!alignmentContexts.containsKey(nsId)) {
+            synchronized (alignmentContexts) {
+              if(!alignmentContexts.containsKey(nsId)) {
+                alignmentContexts.put(nsId, new ClientGSIContext());
+              }
+            }
+          }
```

Comment:
ConcurrentHashMap#putIfAbsent should help with this

Reply:
+1